### PR TITLE
feat: add email notification processor

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

Email support for notifications including different transport methods as well as support for configuring broadcast notification handling. Relates to #23879

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by drodil** [comment]
Tests failing to something not related to this, probably need a rebase

File: plugins/notifications-backend-module-email/package.json
Comment (drodil) [inline plugins/notifications-backend-module-email/package.json:35]: Had to set specific version to prevent installation of newer version of `@aws-sdk/client-sts` that breaks the current functionality.

**Comment by drodil** [comment]
What a fight with aws deps.. Now it's okay to review. @Rugvip / @mareklibra when you have time, thanks!

**Comment by mareklibra** [comment]
@drodil Can you please summarize the difference between this PR and #23879 ?
@ydayagi Can you do the same from your perspective?

It will help to understand what/when/how to proceed towards merge(s).

**Comment by drodil** [comment]
@mareklibra sure, this is a bit more enhanced version of the #23879 as it:

- supports other transports than SMTP (SES and Sendmail)
- configuration follows the same format as the Slack notification processor (#23991)
- broadcast sending is configurable; either recipient emails from config (like it does in #23879), fetching all users, or doing nothing
- cache is used to save the email addresses of users
- it's possible to change the format of the email by extension point

**Comment by drodil** [comment]
Rebased

File: plugins/notifications-backend-module-email/src/processor/transports/sendmail.ts
Comment (freben) [inline plugins/notifications-backend-module-email/src/processor/transports/sendmail.ts:19]: Could make sense to move all of the config reading into these functions as well; that'd alleviate the need for e.g. SendmailTransportConfig as a type to exist at all right? And localizes the functionality better. A bit much code in NotificationsEmailProcessor now that it itself isn't necessarily interested in knowing about, if i understand things right

**Review by freben — APPROVED** [review]
Again random comments, nothing really blocking except the vale warning

File: plugins/notifications-backend-module-email/config.d.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/config.d.ts:22]: Hm, but there are mandatory things you need to configure to use this processor? So I would assume if you add this module, you also need to configure it?

File: plugins/notifications-backend-module-email/config.d.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/config.d.ts:108]: 👍

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts:125]: Hm, I think the cache didn't like the `string[]` for this as it cannot be casted to JsonValue for some reason?

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts:171]: True, might be an issue if there's a lot of emails

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts:125]: Oh it was in the `cache.set` method: `Argument type string[] is not assignable to parameter type JsonValue`

Comment (freben) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts]: What, really? No squigglies in this at least:

```ts
function f(v: JsonValue) {}
const g: string[] = [];
f(g);
```

**Comment by ydayagi** [comment]
> @drodil Can you please summarize the difference between this PR and #23879 ? @ydayagi Can you do the same from your perspective?
> 
> It will help to understand what/when/how to proceed towards merge(s).

@mareklibra @drodil this PR cancels my PR. I will close it.

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts]: Yeah really :D Not sure why, but at least my IDE says it's an issue with `cacheManager.set`, could you try it too?

**Review by mareklibra — APPROVED** [review]

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (freben) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts]: Oh hey CacheManager - that's the old backend-common one - is that the problem, that you have the wrong type on the cache? Should be CacheService

**Comment by drodil** [comment]
Added additional property to throttle the sending, might be useful if using for example SES that has it's limits how many mails can be sent in some period of time.

File: plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts
Comment (drodil) [reply plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts]: Nope, it's `CacheService`. Maybe it's just my IDE messing something up. `tsc:full` completes successfully so it should be ok. I will change this and repush.

**Review by freben — APPROVED** [review]

**Comment by ydayagi** [comment]
> @drodil Can you please summarize the difference between this PR and #23879 ? @ydayagi Can you do the same from your perspective?
> 
> It will help to understand what/when/how to proceed towards merge(s).

this PR is missing a severity threshold.
@drodil can u add that?

**Comment by drodil** [comment]
@ydayagi I think we have to think of the filters for severity and even other things like the topic or scope a bit wider; I think those should be also done for #23991 but I would leave this for later. Maybe another processor function that can be implemented and do the filtering in the `notifications-backend` instead in every processor individually. What do you think about this approach?

Something like:
```
type NotificationProcessorFilters = {
  minSeverity?: NotificationSeverity;
  maxSeverity?: NotificationSeverity;
  excludedTopics?: string[];
  // ... And so on
}

interface NotificationProcessor {
  getNotificationFilters?(): NotificationProcessorFilters;
}
```
And then do the filtering in the backend based on these to determine if the notification should be processed with the specific filter or not. Of course the implementation inside the processor can utilize `Config` or other ways to make the filters.

**Comment by ydayagi** [comment]
> > @ydayagi I think we have to think of the filters for severity and even other things like the topic or scope a bit wider; I think those should be also done for #23991 but I would leave this for later. Maybe another processor function that can be implemented and do the filtering in the `notifications-backend` instead in every processor individually. What do you think about this approach?
> > Something like:
> > ```
> > type NotificationProcessorFilters = {
> >   minSeverity?: NotificationSeverity;
> >   maxSeverity?: NotificationSeverity;
> >   excludedTopics?: string[];
> >   // ... And so on
> > }
> > 
> > interface NotificationProcessor {
> >   getNotificationFilters?(): NotificationProcessorFilters;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And then do the filtering in the backend based on these to determine if the notification should be processed with the specific filter or not. Of course the implementation inside the processor can utilize `Config` or other ways to make the filters.
> 
> I do not know. for our project we need sending entire notification via email when severity is above certain threshold. What about approving both this PR and my PR. we can call my processor smtp-threshold-processor. WDYT? i think severity is a good functionality. the same for putting entire payload in the email body

what u suggested cover all needs

**Comment by drodil** [comment]
Yeah and should also be future-proof. Maybe do that in another PR after merging this to keep things clean

File: plugins/notifications-backend-module-email/src/extensions.ts
Comment (ydayagi) [inline plugins/notifications-backend-module-email/src/extensions.ts:1]: @drodil changing the content of the email requires a user to write code as well as maintain his own module. that makes the user a developer/contributor instead of a user. Let us enable at least some config for that. E.g. emailContnet=all|detailed|short

Comment (drodil) [reply plugins/notifications-backend-module-email/src/extensions.ts:1]: I will not add more features to this PR, all additional work should be done in separate PRs once this is merged

**Comment by drodil** [comment]
@freben maybe this can be merged so we can continue with the additional features in other PRs?

**Comment by freben** [comment]
Sure you already had my approval since before.

**Comment by drodil** [comment]
@freben no I am not, I am not in the maintainer list 😢

**Comment by ydayagi** [comment]
@drodil @freben the link in the email body is for opening the specific notification in the backstage UI?

**Comment by drodil** [comment]
@ydayagi the link is the link included in the notification (if any). I think it needs a bit improvement as it doesn't work with relative links so the app should be added to the link. And also if the notification does not include a link, it should direct to the notifications page. 

If the processor is not installed, it is not active. If it's included in the backend, then the config must be in place or you will end up with error during starting the backend.

**Comment by ydayagi** [comment]
@drodil can we somehow talk?

**Comment by drodil** [comment]
Some improvements done in #24555. 

For talking, I suggest you join the Backstage discord!

**Comment by ydayagi** [comment]
> @ydayagi I think we have to think of the filters for severity and even other things like the topic or scope a bit wider; I think those should be also done for #23991 but I would leave this for later. Maybe another processor function that can be implemented and do the filtering in the `notifications-backend` instead in every processor individually. What do you think about this approach?
> 
> Something like:
> 
> ```
> type NotificationProcessorFilters = {
>   minSeverity?: NotificationSeverity;
>   maxSeverity?: NotificationSeverity;
>   excludedTopics?: string[];
>   // ... And so on
> }
> 
> interface NotificationProcessor {
>   getNotificationFilters?(): NotificationProcessorFilters;
> }
> ```
> 
> And then do the filtering in the backend based on these to determine if the notification should be processed with the specific filter or not. Of course the implementation inside the processor can utilize `Config` or other ways to make the filters.

@drodil have u started working on this item? if not I will be happy to create a PR

**Comment by drodil** [comment]
@ydayagi no, not yet

**Comment by ydayagi** [comment]
@drodil I noticed 2 issues:

1. I get ERR_REQUIRE_ESM because of the throttle library/module
2. config.d.ts has 'transportConfig' key , but the code is looking for 'transport'. should I open a bug?

**Comment by drodil** [comment]
1. Don't know where that is coming from, works just fine in CI and in my local setup.

2. True, it should be just `transport`, no need to open bug, just a PR. I can do it later this week if you don't happen to have time? For some reason there's not really a possibility to do config validation in the tests and they use `transport` config.

**Comment by ydayagi** [comment]
> 1. Don't know where that is coming from, works just fine in CI and in my local setup.
> 2. True, it should be just `transport`, no need to open bug, just a PR. I can do it later this week if you don't happen to have time? For some reason there's not really a possibility to do config validation in the tests and they use `transport` config.

in the CI it is just unit test, is it not? so perhaps nodejs is not used?
which nodejs version do u use?

**Comment by ydayagi** [comment]
@drodil i let a colleague of mine test in his env. same problem with ERR_REQUIRE_ESM. he says it is a known problem with ESM modules

**Comment by drodil** [comment]
@ydayagi when is this error coming up exactly?

**Comment by drodil** [comment]
@ydayagi found this issue: https://github.com/sindresorhus/p-throttle/issues/36

**Comment by ydayagi** [comment]
> @ydayagi when is this error coming up exactly?

The error comes up during startup. The github issue u shared is for the same problem. It is very strange that you do not see this error. I tried the 'import await' workaround. It did not work.

**Comment by ydayagi** [comment]
@drodil 
`
 [backstage]> yarn start-backend:next | tee /tmp/ll
 /home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts:121
       return (
       ^
 
 Error [ERR_REQUIRE_ESM]: require() of ES Module /home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/p-throttle/index.js from /home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts not supported.
 Instead change the require of index.js in /home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts to a dynamic import() which is available in all CommonJS modules.
     at Object.newLoader [as .js] (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:121:7)
     at Object.<anonymous> (/home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/processor/NotificationsEmailProcessor.ts:31:59)
     at Module._compile (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:117:24)
     at Object.newLoader [as .ts] (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:121:7)
     at Object.<anonymous> (/home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/processor/index.ts:25:38)
     at Module._compile (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:117:24)
     at Object.newLoader [as .ts] (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:121:7)
     at Object.<anonymous> (/home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/module.ts:28:20)
     at Module._compile (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:117:24)
     at Object.newLoader [as .ts] (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:121:7)
     at Object.<anonymous> (/home/ydayagi/work/src/github.com/ydayagi/backstage/plugins/notifications-backend-module-email/src/index.ts:29:17)
     at Module._compile (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:117:24)
     at Object.newLoader [as .ts] (/home/ydayagi/work/src/github.com/ydayagi/backstage/node_modules/pirates/lib/index.js:121:7)
     at /home/ydayagi/work/src/github.com/ydayagi/backstage/packages/backend-next/src/index.ts:84:80 {
   code: 'ERR_REQUIRE_ESM'
 }
 
 Node.js v18.15.0`

**Comment by drodil** [comment]
@ydayagi downgrading to version before the ESM change is the solution, see it here: https://github.com/backstage/backstage/pull/24582

**Pull request merged**
